File: ruby-gem/lib/sutra_matrix/calculate.rb
# frozen_string_literal: true

module SutraMatrix
    class InvalidInputError < StandardError; end

    class Calculate
        MATRIX = {
            # [U, I, E, P]
            [0, 0, 0, 0] => 0,
            [1, 0, 0, 0] => 1,
            [0, 1, 0, 0] => 2,
            [1, 1, 0, 0] => 3,
            [0, 0, 1, 0] => 4,
            [1, 0, 1, 0] => 5,
            [0, 1, 1, 0] => 6,
            [1, 1, 1, 0] => 7,
            [0, 0, 0, 1] => 8,
            [1, 0, 0, 1] => 9,
            [0, 1, 0, 1] => 10,
            [1, 1, 0, 1] => 11,
            [0, 0, 1, 1] => 12,
            [1, 0, 1, 1] => 13,
            [0, 1, 1, 1] => 14,
            [1, 1, 1, 1] => 15
          }.freeze

        def self.score(u, i, e, p)
        raise InvalidInputError, "U, I, E, P should be binary (0 or 1)" unless [u, i, e, p].all? { |x| [0, 1].include?(x) }

        MATRIX[[u, i, e, p]]
        end
    end
end

File: node-package/lib/calculate.js
class SutraMatrix {
    static MATRIX = new Map([
      [[0, 0, 0, 0].toString(), 0],  [[1, 0, 0, 0].toString(), 1],
      [[0, 1, 0, 0].toString(), 2],  [[1, 1, 0, 0].toString(), 3],
      [[0, 0, 1, 0].toString(), 4],  [[1, 0, 1, 0].toString(), 5],
      [[0, 1, 1, 0].toString(), 6],  [[1, 1, 1, 0].toString(), 7],
      [[0, 0, 0, 1].toString(), 8],  [[1, 0, 0, 1].toString(), 9],
      [[0, 1, 0, 1].toString(), 10], [[1, 1, 0, 1].toString(), 11],
      [[0, 0, 1, 1].toString(), 12], [[1, 0, 1, 1].toString(), 13],
      [[0, 1, 1, 1].toString(), 14], [[1, 1, 1, 1].toString(), 15]
    ]);
  
    static score(u, i, e, p) {
      if (![u, i, e, p].every(x => x === 0 || x === 1)) {
        throw new Error("U, I, E, P should be binary values (0 or 1)");
      }
      return this.MATRIX.get([u, i, e, p].toString());
    }
  }
  
  module.exports = SutraMatrix;
  